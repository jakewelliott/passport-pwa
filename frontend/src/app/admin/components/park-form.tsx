import { useField } from '@/hooks/useField';
import type { DatabaseEntry, Park as ParkT } from '@/types';
import { PARK_ICONS, type ParkIcon } from '@/types/icons';
import { useForm } from '@tanstack/react-form';
import { z } from 'zod';

type Park = Omit<ParkT, keyof DatabaseEntry>;

// this is somewhat expensive
const isParkIcon = (obj: any): obj is ParkIcon => {
    return PARK_ICONS.some(
        (icon) => icon.iconName === obj.iconName && icon.tooltip === obj.tooltip && icon.extraText === obj.extraText,
    );
};

const parkIconSchema = z.custom<ParkIcon>(isParkIcon);

// TODO: make this match the database AND readability stuff
// i.e. parkName should have "State Park" at the end
// coordinates should be to xyz decmial places

// NOTE: some of the fields that are stored internally in
// different tables and joined in the request will need separate
// fields in the form

// ADAM: jake i just am now seeing ur edit-park-modal.tsx
// so i will reuse the UI u made there for addresses, icons, photos, etc.
const parkSchema = z.object({
    parkName: z.string().min(1),
    coordinates: z.object({
        latitude: z.number(),
        longitude: z.number(),
        inaccuracyRadius: z.number(),
    }),
    phone: z.number(),
    email: z.string().email(),
    establishedYear: z.string().regex(/^(19|20)\d{2}$/, { message: 'Invalid year' }),
    landmark: z.string().min(1),
    youCanFind: z.string().min(1),
    trails: z.string().min(1),
    website: z.string().url(),
    addresses: z.array(
        z.object({
            title: z.string().min(1),
            addressLineOne: z.string().min(1),
            addressLineTwo: z.string().min(1),
            city: z.string().min(1),
            state: z.string().min(1),
            zipcode: z.number().min(1),
        }),
    ),
    icons: z.array(parkIconSchema),
    photos: z.array(
        z.object({
            photoPath: z.string(),
            alt: z.string(),
        }),
    ),
    abbreviation: z.string().min(1),
    stampImage: z.string().min(1),
    accesses: z.string().min(1),
});

export const EditParkForm = ({ park, refetch }: { park: Park; refetch: () => void }) => {
    const form = useForm({
        defaultValues: {
            ...park,
        },
        onSubmit: async (values) => {
            // Submit to API
            console.log(values);
        },
        validators: {
            onBlur: parkSchema,
            onSubmit: parkSchema,
        },
    });

    const Field = useField(form);

    return (
        <div className='flex flex-col gap-4'>
            <Field name='parkName' label='Park Name' />
            <Field name='coordinates' label='Coordinates' />
            <Field name='phone' label='Phone' />
            <Field name='email' label='Email' />
            <Field name='establishedYear' label='Established Year' />
            <Field name='landmark' label='Landmark' />
            <Field name='youCanFind' label='You Can Find' />
            <Field name='trails' label='Trails' />
            <Field name='website' label='Website' />
            {/* TODO: make interfaces for these fields */}
            {/* <Field name='addresses' label='Addresses' /> */}
            {/* <Field name='icons' label='Icons' /> */}
            {/* <Field name='photos' label='Photos' /> */}
            <Field name='abbreviation' label='Abbreviation' />
            <Field name='stampImage' label='Stamp Image' />
            <Field name='accesses' label='Accesses' />
        </div>
    );
};
