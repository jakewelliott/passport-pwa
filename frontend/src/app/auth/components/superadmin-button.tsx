import RoundedButton from "@/components/rounded-button";
import { dbg, PRODUCTION } from "@/lib/debug";
import { useLogin } from "@/hooks/auth/useLogin";

const username = process.env.ADMIN_USER || '';
const password = process.env.ADMIN_PASS || '';

if (!username || !password) {
	dbg('MISC', 'SuperAdminButton', 'ADMIN_USER or ADMIN_PASS env var not set');
}

const credentials = { username, password };

export const SuperAdminButton = () => {

	const login = useLogin();
	if (PRODUCTION) return null;

	const handleClick = async (e: React.FormEvent) => {
		e.preventDefault();
		dbg('AUTH', 'SUPER ADMIN LOGIN', credentials);
		login.mutate(credentials);
	};

	return (
		<button type="button" onClick={handleClick} data-testid="super-admin-button">
			<RoundedButton title="SUPERADMIN" />
		</button>
	)
};
