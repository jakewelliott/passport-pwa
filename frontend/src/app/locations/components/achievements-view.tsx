import type { Park, ParkActivity } from '@/lib/mock/types';
import { FaRegCheckSquare, FaRegSquare, FaStamp } from 'react-icons/fa';

interface AchievementsViewProps {
  parkActivity: ParkActivity;
  park: Park;
}

const AchievementsView = ({ parkActivity, park }: AchievementsViewProps) => {
  return (
    <div className='flex flex-col gap-3' data-testid='achievements-view'>
      {parkActivity.stampCollectedAt !== undefined && (
        <div className='top-0 flex'>
          <FaStamp size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
          <p>
            {parkActivity.stampCollectedAt
              ? `Stamp collected ${parkActivity.stampCollectedAt}`
              : 'Stamp not yet collected'}
          </p>
        </div>
      )}
      {
        park.bucketListItems &&
        park.bucketListItems.length > 0 && (
          <div className='top-0 flex'>
            {parkActivity.completedBucketListItems[0]?.id ? (
              <FaRegCheckSquare size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
            ) : (
              <FaRegSquare size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
            )}
            <p data-testid={'BLI'}>
              Bucket List Item:
              <br />
              {park.bucketListItems[0].task}
            </p>
          </div>
        )}
    </div>
  );
};

export default AchievementsView;
