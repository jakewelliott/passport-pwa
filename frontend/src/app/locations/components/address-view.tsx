import type { Address, Park } from '@/types';
import { useState } from 'react';
import { FiNavigation } from 'react-icons/fi';


const ParkAddress = ({ address }: { address: Address }) => {
	return (
		<div>
			<p>{address.title}</p>
			<p>{address.addressLineOne}</p>
			{address.addressLineTwo && <p>{address.addressLineTwo}</p>}
			<p>{address.city}, {address.state} {address.zipcode}</p>
		</div>
	);
};

export const AddressView = ({ park }: { park: Park }) => {

	const [showAll, setShowAll] = useState(false);
	const showButton = park.addresses && park.addresses.length > 1;

	const toggle = () => {
		setShowAll(!showAll);
	};

	const OtherAddresses = () =>
		park.addresses.slice(1).map((address) => (
			<ParkAddress key={address.title} address={address} />
		))


	const Button = () =>
		<button type='button' onClick={toggle} className='ml-4 inline-flex items-center'>
			{showAll ? 'Show Less' : 'Show More'}
		</button>

	return (
		<div className='top-0 flex' key={park.parkName}>
			<FiNavigation size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
			<ParkAddress address={park.addresses[0]} />
			{showAll && <OtherAddresses />}
			{showButton && <Button />}
		</div>
	);
};
