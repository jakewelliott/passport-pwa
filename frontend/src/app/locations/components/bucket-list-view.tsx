import { useBucketList } from '@/hooks/queries/useBucketList';
import { useStamp } from '@/hooks/queries/useStamps';
import DateHelper from '@/lib/date-helper';
import type { BucketListItem, CollectedStamp, Park } from '@/lib/mock/types';
import { FaRegCheckSquare, FaRegSquare, FaStamp } from 'react-icons/fa';

const BucketListItemView = ({ bucketListItem }: { bucketListItem: BucketListItem }) => {
	const isCompleted = bucketListItem.completedAt && !bucketListItem.deleted;
	const Checkbox = isCompleted ? FaRegCheckSquare : FaRegSquare;

	return (
		<div>
			<Checkbox size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
			<p data-testid={'BLI'}>
				Bucket List Item:
				<br />
				{bucketListItem.task}
			</p>
		</div>
	);
};

const StampView = ({ stamp }: { stamp: CollectedStamp | undefined }) => {
	return (
		<>
			<FaStamp size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
			<p>
				{stamp
					? `Stamp collected on ${DateHelper.stringify(new Date(stamp.createdAt)).replace(',', ' at')}`
					: 'Stamp not yet collected'}
			</p>
		</>
	);
};

export const BucketListView = ({ park }: { park: Park }) => {
	const { data: items, isLoading: itemsLoading } = useBucketList(park.id);
	const { data: stamp, isLoading: stampLoading } = useStamp(park.abbreviation);

	// TODO: replace with a placeholder
	if (itemsLoading || stampLoading) return null;

	return (
		<div className='flex flex-col gap-3' data-testid='achievements-view'>
			<div className='top-0 flex'>
				<StampView stamp={stamp} />
			</div>
			<div className='top-0 flex'>
				{items.map((item: BucketListItem) => (
					<BucketListItemView key={item.id} bucketListItem={item} />
				))}
			</div>
		</div>
	);
};
