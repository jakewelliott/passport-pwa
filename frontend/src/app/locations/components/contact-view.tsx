import type { Park } from '@/types';
import { FaRegEnvelope } from 'react-icons/fa';
import { FiNavigation, FiPhone } from 'react-icons/fi';

export const ContactView = ({ park }: { park: Park }) => {
    return (
        <div className='mx-4 flex flex-col gap-3' data-testid='location-contact'>
            {/* Coordinates */}
            <div className='top-0 flex'>
                <FiNavigation size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
                <p>
                    GPS: {park.coordinates.latitude.toFixed(4)}, {park.coordinates.longitude.toFixed(4)}
                </p>
            </div>
            {/* Phone */}
            <div className='top-0 flex'>
                <FiPhone size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
                <p>{String(park.phone).replace(/(\d{3})(\d{3})(\d{4})/, '($1) $2-$3')}</p>
            </div>
            {/* Email */}
            <div className='top-0 flex'>
                <FaRegEnvelope size={'17px'} strokeWidth={3} style={{ paddingRight: '5px', paddingTop: '5px' }} />
                <p>{park.email}</p>
            </div>
        </div>
    );
};
