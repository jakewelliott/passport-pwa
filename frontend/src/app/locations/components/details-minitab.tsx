import type { Park } from '@/types';
import React from 'react';

export const DetailsMiniTab = ({ park }: { park: Park }) => {
	const renderTrails = (trails: string) => {
		return trails
			.split('\n')
			.filter((line) => line.trim() !== '') // Remove empty lines
			.map((line, index, array) => (
				<React.Fragment key={line}>
					{line}
					{index < array.length - 1 && <br />} {/* Add <br /> only if it's not the last line */}
				</React.Fragment>
			));
	};

	return (
		<div className='mt-6 mb-6 flex flex-col'>
			<p className='pr-6 pl-6'>
				{park.establishedYear && (
					<>
						<span className='green-text'>Established: </span>
						{park.establishedYear}
						<br />
						<br />
					</>
				)}
				{park.landmark && (
					<>
						<span className='green-text'>Landmark: </span>
						{park.landmark}
						<br />
						<br />
					</>
				)}
				{park.youCanFind && (
					<>
						<span className='green-text'>You can find... </span>
						{park.youCanFind}
						<br />
						<br />
					</>
				)}
				{park.trails && (
					<>
						<span className='green-text'>Trails: </span>
						<br />
						{renderTrails(park.trails)}
						<br />
						<br />
					</>
				)}
			</p>
			<div data-testid='icon-scroll-container' className='icon-scroll-container overflow-x-auto'>
				<div className='inline-flex gap-6 px-6'>
					{park.icons?.map((icon, index) => (
						<img
							src={`/icons/park/${icon.iconName}.svg`}
							width={55}
							height={55}
							key={icon.iconName}
							alt={`Park icon ${index + 1}`}
						/>
					))}
					<div className='w-px flex-shrink-0' />
				</div>
			</div>
		</div>
	);
};
