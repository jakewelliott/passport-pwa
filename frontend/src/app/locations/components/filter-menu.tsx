import { FaFilter } from 'react-icons/fa';

interface FilterMenuProps {
    searchQuery: string;
    setSearchQuery: (value: string) => void;
    setIsFilterModalOpen: (value: boolean) => void;
}

export const FilterMenu = ({ searchQuery, setSearchQuery, setIsFilterModalOpen }: FilterMenuProps) => (
    <div className='relative flex gap-2'>
        <div className='relative flex-1'>
            <input
                type='text'
                placeholder='Search parks...'
                value={searchQuery}
                onChange={(e) => setSearchQuery(e.target.value)}
                className='w-full rounded-lg border border-system-gray bg-supporting-lightblue p-3 pr-7 focus:border-secondary-darkteal focus:outline-none focus:ring-1 focus:ring-secondary-darkteal focus:ring-opacity-100'
            />
            {searchQuery && (
                <button
                    type='button'
                    onClick={() => setSearchQuery('')}
                    className='-translate-y-1/2 absolute top-1/2 right-3 transform bg-supporting-lightblue text-system-gray hover:text-secondary-darkteal'
                    aria-label='Clear search'
                >
                    &times;
                </button>
            )}
        </div>
        <button
            type='button'
            onClick={() => setIsFilterModalOpen(true)}
            className={
                'w-auto rounded-lg border border-system-gray bg-system-white px-3 py-2 transition-all duration-200'
            }
        >
            <FaFilter />
        </button>
    </div>
);
