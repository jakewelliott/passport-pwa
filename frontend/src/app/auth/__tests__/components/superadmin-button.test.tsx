import { SuperAdminButton } from '@/app/auth/components/superadmin-button';
import { renderWithClient } from '@/lib/test-wrapper';
import { screen } from '@testing-library/react';
import { describe, expect, it } from 'vitest';

describe('SuperAdmin Button', () => {
	it('renders the button', () => {
		renderWithClient(<SuperAdminButton />);
		expect(screen.getByText('SUPERADMIN')).toBeInTheDocument();
	});
});