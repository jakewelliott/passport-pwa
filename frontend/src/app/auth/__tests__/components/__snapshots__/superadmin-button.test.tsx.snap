// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SuperAdmin Button > matches snapshot 1`] = `
<div>
  <button
    class="bg-secondary_darkteal flex cursor-pointer select-none items-center justify-center text-center p-4 rounded-full"
    style="min-width: 128px;"
    type="button"
  >
    <p
      class="text-system_white button-text"
    >
      SUPERADMIN
    </p>
  </button>
</div>
`;
